node/impls.go

```diff
@@ -161,108 +162,29 @@ func (l agreementLedger) EnsureValidatedBlock(ve agreement.ValidatedBlock, c agr
 }
 
 // EnsureDigest implements agreement.LedgerWriter.EnsureDigest.
-// TODO: Get rid of EnsureDigest -- instead the ledger should expose what blocks it's waiting on, and a separate service should fetch them and call EnsureBlock
-// should "retry until cert matches" logic live here or in the abstract fetcher?
 func (l agreementLedger) EnsureDigest(cert agreement.Certificate, quit chan struct{}, verifier *agreement.AsyncVoteVerifier) {
-	round := cert.Round
-	blockHash := bookkeeping.BlockHash(cert.Proposal.BlockDigest) // semantic digest (i.e., hash of the block header), not byte-for-byte digest
-	logging.Base().Debug("consensus was reached on a block we don't have yet: ", blockHash)
-	for {
-		// Ask the fetcher to get the block somehow
-		block, fetchedCert, err := l.FetchBlockByDigest(round, quit)
-		if err != nil {
-			select {
-			case <-quit:
-				logging.Base().Debugf("EnsureDigest was asked to quit before we could acquire the block")
-				return
-			default:
-			}
-			logging.Base().Panicf("EnsureDigest could not acquire block, fetcher errored out: %v", err)
-		}
-
-		if block.Hash() == blockHash && block.ContentsMatchHeader() {
-			l.EnsureBlock(block, cert)
-			return
-		}
-		// Otherwise, fetcher gave us the wrong block
-		logging.Base().Warnf("fetcher gave us bad/wrong block (for round %d): fetched hash %v; want hash %v", round, block.Hash(), blockHash)
-
-		// As a failsafe, if the cert we fetched is valid but for the wrong block, panic as loudly as possible
-		if cert.Round == fetchedCert.Round &&
-			cert.Proposal.BlockDigest != fetchedCert.Proposal.BlockDigest &&
-			fetchedCert.Authenticate(block, l, verifier) == nil {
-			s := "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s += "!!!!!!!!!! FORK DETECTED !!!!!!!!!!!\n"
-			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s += "EnsureDigest called with a cert authenticating block with hash %v.\n"
-			s += "We fetched a valid cert authenticating a different block, %v. This indicates a fork.\n\n"
-			s += "Cert from our agreement service:\n%#v\n\n"
-			s += "Cert from the fetcher:\n%#v\n\n"
-			s += "Block from the fetcher:\n%#v\n\n"
-			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s += "!!!!!!!!!! FORK DETECTED !!!!!!!!!!!\n"
-			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s = fmt.Sprintf(s, cert.Proposal.BlockDigest, fetchedCert.Proposal.BlockDigest, cert, fetchedCert, block)
-			fmt.Println(s)
-			logging.Base().Error(s)
-		}
+	waitForRound := l.Wait(cert.Round)
+	// clear out the pending certificates ( if any )
+	select {
+	case pendingCert := <-l.UnmatchedPendingCertificates:
```

Comment:
Does this currently ever happen?

Reply:
It could. The is a very low probability that the catchup is at a stage where it would try to catch up after no new block was received by the agreement service. if that happens, we could have ended up the first unmatched pending certificate to be fulfilled by a "regular" catchup. at that point, we have a "stale" item in the "UnmatchedPendingCertificates". 
We'd probably do better by addressing this edge-case in both EnsureDigest as well as in catchup service.